presto-main/src/main/java/com/facebook/presto/server/remotetask/TaskInfoFetcher.java

```diff
@@ -136,6 +141,18 @@ private synchronized void stop()
         }
     }
 
+    public void addFinalTaskInfoListener(StateChangeListener<TaskInfo> stateChangeListener)
+    {
+        AtomicBoolean done = new AtomicBoolean();
+        StateChangeListener<Optional<TaskInfo>> fireOnceStateChangeListener = finalTaskInfo -> {
+            if (finalTaskInfo.isPresent() && done.compareAndSet(false, true)) {
+                stateChangeListener.stateChanged(finalTaskInfo.get());
+            }
+        };
+        finalTaskInfo.addStateChangeListener(fireOnceStateChangeListener);
+        fireOnceStateChangeListener.stateChanged(finalTaskInfo.get());
```

Comment:
Why do we need to call `stateChanged` here? If the state is terminal, `addStateChangeListener` would just fire the listener.

Reply:
I also noticed that. For some reason we don't register the final state as terminal in other places as well (e.g. QueryExecution, StageExecution, etc.)

Reply:
@nezihyigitbasi, it is a design flaw in the `StateMachine` code, which will take a lot of care to fix.  The root issue is that when a listener is registered, it only sees future state changes, but it is possible that the final state has already been set.  The fix we need is to always fire the current state to when a state change listener is added, but I am concerned that is might break some existing code, so I have never taken up the effort to make this change.